src/engine/renderer/glsl_source/reliefMapping_fp.glsl

```diff
@@ -29,21 +29,169 @@ uniform sampler2D	u_NormalMap;
 uniform vec3        u_NormalScale;
 #endif // r_normalMapping
 
-#if defined(USE_PARALLAX_MAPPING)
-#if !defined(USE_HEIGHTMAP_IN_NORMALMAP)
+#if (defined(USE_PARALLAX_MAPPING) && !defined(USE_HEIGHTMAP_IN_NORMALMAP)) || defined(USE_NORMALMAP_FROM_HEIGHTMAP)
 uniform sampler2D	u_HeightMap;
-#endif // !USE_HEIGHTMAP_IN_NORMALMAP
+#endif // (USE_PARALLAX_MAPPING && !USE_HEIGHTMAP_IN_NORMALMAP) || USE_NORMALMAP_FROM_HEIGHTMAP
+
+#if defined(USE_PARALLAX_MAPPING)
 uniform float       u_ParallaxDepthScale;
 uniform float       u_ParallaxOffsetBias;
 #endif // USE_PARALLAX_MAPPING
 
+#if defined(USE_NORMALMAP_FROM_HEIGHTMAP)
+vec3 NormalFromHeightMap(vec2 texNormal)
+{
+	vec3 normal;
+
+	// major inspiration was this post by jollyjeffers:
+	//   https://www.gamedev.net/forums/topic/475213-generate-normal-map-from-heightmap-algorithm/#post_4117038
+	// and this Wikipedia article:
+	//   https://en.wikipedia.org/wiki/Sobel_operator
+	//
+	// Various people recommends doing abs() on sample read in case of float image format:
+	//   http://www.catalinzima.com/2008/01/converting-displacement-maps-into-normal-maps/
+	//   https://community.khronos.org/t/heightmap-to-normalmap/58862
+	// Or report issues by not doing it:
+	//   https://gamedev.stackexchange.com/questions/165575/calculating-normal-map-from-height-map-using-sobel-operator
+	// Other people say texture2D clamps it but it's safer to do this.
+
+	// Get texel size
+	vec2 texelSize = 1.0 / vec2(textureSize(u_HeightMap, 0));
+
+#if defined(r_sobelFiltering)
+	// Useful things to know:
+	//
+	// * It's required to get height samples (S) surrounding the current texel (T) to do a sobel filter:
+	//
+	//   S S S
+	//   S T S
+	//   S S S
+	//
+	// * Sobel X kernel is:
+	//
+	//   [ 1  0  -1 ]
+	//   [ 2  0  -2 ]
+	//   [ 1  0  -1 ]
+	//
+	//   See https://en.wikipedia.org/wiki/Sobel_operator#Formulation
+	//
+	// * Sobel Y kernel is:
+	//
+	//   [  1  2  1 ]
+	//   [  0  0  0 ]
+	//   [ -1 -2 -1 ]
+	//
+	//   Which is the rotated sobel X kernel.
+	//
+	// * Tangent space normals are +Z.
+
+	ivec3 offsets;
+	mat3 xCoords;
+	mat3 yCoords;
+	mat3 heights;
+	mat3 sobel;
+
+	// Components will be computed by accumulating values.
+	normal = vec3(0.0, 0.0, 0.0);
+
+	// Set offsets:
+	offsets = ivec3(-1, 0, 1);
+
+	// Set sobel X kernel (beware, line is column with this notation):
+	sobel[0] = vec3( 1.0,  2.0,  1.0);
+	sobel[1] = vec3( 0.0,  0.0,  0.0);
+	sobel[2] = vec3(-1.0, -2.0, -1.0);
+
+	for (int i = 0; i < 3; i++)
+	{
+		for (int j = 0; j < 3; j++)
+		{
+			if (i != 1 && j != 1)
```

Comment:
This condition looks wrong

Reply:
one matrix is:

    N N N
    0 0 0
    N N N

while other is:

    N 0 N
    N 0 N
    N 0 N

`mat[1][1]` is know to be zero in all cases, because only surrounding pixels are read, not the pixel itself.

Reply:
note that I'm not sure to understand everything… but other implementations I read just did similar thing (they basically unrolled loops, ignoring the useless unrolled step).

Reply:
This also ignores [0][1] and [2][1] and [1][0] and [2][0]

Reply:
Hmm, I think I wanted to do `!( i == 1 && j == 1)`
…which must be equivalent to `(i != 1 || j !=1)`, right.